bunsen/bunsen-core-r4/src/main/java/com/cerner/bunsen/definitions/r4/R4StructureDefinitions.java

```diff
@@ -38,8 +39,11 @@ public FhirConversionSupport conversionSupport() {
 
   @Override
   protected IStructureDefinition getStructureDefinition(String resourceUrl) {
-    return new StructureDefinitionWrapper(
-        (StructureDefinition) context.getValidationSupport().fetchStructureDefinition(resourceUrl));
+    IBaseResource baseResource =
+        context.getValidationSupport().fetchStructureDefinition(resourceUrl);
+    return baseResource == null
```

Comment:
I see that in many cases that call this method, you are catching the `null` case and throwing a `RuntimeException` for it; is there a reason not to throw such exceptions here? I mean is there a case where we can do anything else in the `null` case?

Reply:
What is the contract of `getStructureDefinition`; I mean is `null` an expected/possible return value? Asking because now after looking a little bit closer, I am more concerned if `getStructureDefinition` can return `null`. For example [here](https://github.com/google/fhir-data-pipes/blob/498d7c1b336bf77915fc39ca973b4231b594bf28/bunsen/bunsen-core/src/main/java/com/cerner/bunsen/definitions/StructureDefinitions.java#L316) we don't handle `null` and potentially have an NPE.

So here is my suggestion: Either (i) establish that `null` is a valid return value of `getStructureDefinition` and then handle `null` everywhere this is called; OR (ii) add @NotNull as part of its return contract (in the base interface) and in both R4 and Stu3 implementations, check that we do not return `null` (e.g., throw an exception).

Reply:
But I think the question still remains, is returning `null` an expected return value of `getStructureDefinition`? To clarify further, the example you mentioned earlier from [here](https://github.com/google/fhir-data-pipes/blob/498d7c1b336bf77915fc39ca973b4231b594bf28/bunsen/bunsen-core/src/main/java/com/cerner/bunsen/definitions/StructureDefinitions.java#L621), is not really the return value of `getStructureDefinition` but `getDefinition`. In other words, if `getDefinition` return `null` [here](https://github.com/google/fhir-data-pipes/blob/498d7c1b336bf77915fc39ca973b4231b594bf28/bunsen/bunsen-core/src/main/java/com/cerner/bunsen/definitions/StructureDefinitions.java#L338) it seems to me that it is assumed the type is `BackboneElement` or `Element`. Now if we change the contract of `getStructureDefinition` we are breaking that assumption.

So to summarize, I suggest that we never return `null` from `getStructureDefinition` and clean up all the exceptions that are added (move them to a single place in `getStructureDefinition`), unless if there is a legitimate case for which I appreciate if a unit-test is added demonstrating that `getStructureDefinition` returning `null` is a legitimate case.

Reply:
Ok, makes sense. In the latest commit, I have added the **@Nonnull** annotation and a `@throws IllegalArgumentException` to the method contract in both the interface and implementations. Though the `@throws` declaration was not mandatory (as its not a checked exception), I have added for the sake of clarity.